Transactional Indexes

sergeybykov opened the pull request.

TedHartMS committed:
WIP on Transactional Indexes

TedHartMS committed:
Make Transactional Lookups work

TedHartMS committed:
Player tests for Transactional (and add some update-testing as well)

TedHartMS committed:
For Transactional, defer UpdateBeforeImage to be before the updateFunc execution, so there are no stale leftovers in the grain if there is a rollback; Add Transactional tests for MultiIndex and MultiInterface; add tests for index updates; fix a bug in non-SingleBucket index updates (add IndexUpdateMode to IMemberUpdate so MemberUpdateOverriddenOperation will carry it through to HashIndexBucketUtils) and another in GenerateMemberUpdates (consider an update of unique to be two unique operations); Move ApplyIndexUpdatesEagerly and UpdateBeforeImages from IndexedStateBase to WorkflowIndexedStateBase

TedHartMS committed:
Switch to Grain Lifecycle (instead of call-throughs from OnActivate/DeactivateAsync); fix test to remove conflict on IPlayerChain1Grain giving inconsistent results

TedHartMS committed:
Add StateName to IndexedStateAttribute; Move ITransactionalState into ITransactionalIndexState; undo some mistaken #if-out of TotalIndex tests

TedHartMS committed:
Use Microsoft.Extensions.Options for OptionsBuilder; temporarily disable CosmosDB in Indexing pending a similar change in CosmosDB

TedHartMS committed:
Skip BaseIndexingFixture CosmosDB read from config file if no DBname; remove _Wf extension from tests; add a couple comments and TODOs

TedHartMS committed:
Txn multi-hashbucketgrain deadlock mitigation; remove workflow-specific naming in tests; clean up some TODOs

TedHartMS committed:
DSMI infrastructure for transactions (via TransactionalStateStorageProviderWrapper); however DSMI is still #ifdef'd out until Orleans 2.3.0 is released and Orleans.CosmosDB is updated to it (due to conflicting definitions of OptionsBuilder in previous Orleans vs. newer MS.Extensions.Options)

galvesribeiro commented:
CosmosDB provider has a pending PR for custom indexes on grain state that I’m waiting for 2.3.0 to merge it. Please hold on this until that.

TedHartMS committed:
Add transaction benchmarks and commit/rollback Player tests

TedHartMS committed:
Add MaxHashBuckets to Indexing

TedHartMS committed:
Update SportsTeamIndexing sample and IndexingFacet.md to latest interfaces

galvesribeiro commented:
FYI, CosmosDB providers were published.

TedHartMS committed:
Update to latest CosmosDB

TedHartMS committed:
Make Active non-PerSilo illegal and remove tests; rename Player tests to reflect the type of grain interface and remove no-longer-valid tests/files; remove "#if ALLOW_FT_ACTIVE" blocks; update doc with additional partitioning info

TedHartMS committed:
Rename individual Player tests to be more informative; Remove all no-longer-invalid Active Index specifications from Multi* tests; remove no-longer-valid Active Index PerKey/SingleBucket implementation files; remove all "#if ALLOW_FT_ACTIVE" blocks as that is no longer valid; add PlayerChain tests to ensure all GetNextBucket overrides are exercised

TedHartMS committed:
- Support multiple grains implementing an indexed grain interface, update TransactionalIndexedState's invariant sequence to support this, add a test for it, and update the documentation.
- Remove some lingering "Workflow" naming
- Fix some broken links in the document
- Remove unused grainClassType parameter from IndexFactory.RegisterIndexWorkflowQueues

TedHartMS committed:
Add IndexingInternals.md and some code comments

TedHartMS committed:
- Move GrainService registration to Configuration time and remove Silo.AddGrainService
- Modifications to remove dependencies on, and then reverse, the changes from internal -> public in GrainTypeResolver and GrainClassData and related methods from Silo.cs
- Add isHandler param to CreateIndexWorkflowQueuePrimaryKey for debuggability

TedHartMS committed:
- Add tests and fixes for FT grain reactivation with and without use of ReincarnatedWorkflowQueue
- add IInjectableCode
- remove unused Silo from IndexWorkflowQueue(State|Entry)

philbe committed:
Update IndexingInternals.md

Fix typos and the like

TedHartMS committed:
Minor changes per Phil's feedback on IndexingInternals.md
